src/pd/leader.rs

```diff
@@ -1,51 +1,222 @@
+use super::{Error, PdFuture, PdTimestamp, Result, PD_REQUEST_HISTOGRAM_VEC, REQUEST_TIMEOUT};
+use tokio_core::reactor::{Core, Handle as OtherHandle};
 use util::security::SecurityManager;
+use util::time::duration_to_sec;
 use util::timer::GLOBAL_TIMER_HANDLE;
-use util::{Either, HandyRwLock};
-
-macro_rules! box_err {
-    ($e:expr) => ({
-        use std::error::Error;
-        let e: Box<Error + Sync + Send> = format!("[{}:{}]: {}", file!(), line!(),  $e).into();
-        e.into()
-    });
-    ($f:tt, $($arg:expr),+) => ({
-        box_err!(format!($f, $($arg),+))
-    });
+use util::HandyRwLock;
+
+macro_rules! request {
+    ($cluster_id:expr, $type:ty) => {{
+        let mut request = <$type>::new();
+        let mut header = ::kvproto::pdpb::RequestHeader::new();
+        header.set_cluster_id($cluster_id);
+        request.set_header(header);
+        request
+    }};
+}
+
+type TsoChannel = oneshot::Sender<PdTimestamp>;
+
+type PdRequest = Box<Future<Item = (), Error = ()> + Send>;
+
+pub enum PdTask {
+    TsoInit,
+    TsoRequest,
+    TsoResponse(Vec<oneshot::Sender<PdTimestamp>>, TsoResponse),
+    Request(PdRequest),
+}
+
+struct PdReactor {
+    task_tx: Option<UnboundedSender<Option<PdTask>>>,
+    tso_tx: UnboundedSender<TsoRequest>,
+    tso_rx: Option<UnboundedReceiver<TsoRequest>>,
+
+    handle: Option<JoinHandle<()>>,
+    tso_pending: Option<Vec<TsoChannel>>,
+    tso_buffer: Option<Vec<TsoChannel>>,
+    tso_batch: Vec<TsoChannel>,
+}
+
+impl Drop for PdReactor {
+    fn drop(&mut self) {
+        if let Some(handle) = self.handle.take() {
+            handle.join().unwrap();
+        }
+    }
+}
+
+impl PdReactor {
+    fn new() -> PdReactor {
+        let (tso_tx, tso_rx) = unbounded();
+        PdReactor {
+            task_tx: None,
+            tso_tx,
+            tso_rx: Some(tso_rx),
+            handle: None,
+            tso_buffer: Some(Vec::with_capacity(8)),
+            tso_batch: Vec::with_capacity(8),
+            tso_pending: None,
+        }
+    }
+
+    fn start(&mut self, client: Arc<RwLock<LeaderClient>>) {
+        if self.handle.is_none() {
+            info!("starting pd reactor thread");
+            let (task_tx, task_rx) = unbounded();
+            task_tx.unbounded_send(Some(PdTask::TsoInit)).unwrap();
+            self.task_tx = Some(task_tx);
+            self.handle = Some(
+                thread::Builder::new()
+                    .name("dispatcher thread".to_owned())
+                    .spawn(move || Self::poll(&client, task_rx))
+                    .unwrap(),
+            )
+        } else {
+            warn!("tso sender and receiver are stale, refreshing..");
+            let (tso_tx, tso_rx) = unbounded();
+            self.tso_tx = tso_tx;
+            self.tso_rx = Some(tso_rx);
+            self.schedule(PdTask::TsoInit);
+        }
+    }
+
+    fn schedule(&self, task: PdTask) {
+        self.task_tx
+            .as_ref()
+            .unwrap()
```

Comment:
You can use `.expect()` to give some context to errors when they happen. It makes it a bit easier to understand.